asab/library/providers/git.py

```diff
@@ -0,0 +1,127 @@
+import tempfile
+import logging
+import os
+import shutil
+
+import pygit2
+
+from .filesystem import FileSystemLibraryProvider
+from ...config import Config
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+
+class GitLibraryProvider(FileSystemLibraryProvider):
+	"""
+	Read-only git provider to read from remote repository.
+	It clones a remote git repository to a temporary directory and then uses the
+	FileSystemLibraryProvider to read the files.
+	To read from local git repository, please use FileSystemProvider.
+
+	Configuration:
+	(Use either deploytoken, publickey+privatekey for SSH option, or username and password and HTTP access.)
+	```
+	[library]
+	providers=git+<URL or deploy token>
+
+	[library.GitProvider]
+	publickey=<absolute path to file>
+	privatekey=<absolute path to file>
+	username=johnsmith
+	password=secretpassword
+	```
+	"""
+	def __init__(self, library, path):
+		self.URL = path[4:]
+		self.Callbacks = pygit2.RemoteCallbacks(get_git_credentials(self.URL))
+		self.RepoPath = tempfile.TemporaryDirectory().name
+		self.GitRepository = clone_repository(self.URL, self.RepoPath, self.Callbacks)
+		self._check_remote()
+		super().__init__(library, self.RepoPath)
+
+		from ...proactor import Module
+		self.App.add_module(Module)
+		self.ProactorService = self.App.get_service("asab.ProactorService")
+
+		self.App.PubSub.subscribe("Application.tick/60!", self._periodic_pull)
+
+
+	async def finalize(self, app):
+		clean_path(self.RepoPath)
```

Comment:
I guess you can keep the repository on the drive ... and do "update" on start of the service.

Reply:
I can do this. I didn't want to leave a mess behind when the application exits for good. Also... when keeping the repository persistent on the drive - I would make the location configurable - so the user knows where to look for it (it is not temporary anymore), but then it is much more accessible... And it might be changed locally, and these changes might cause merge conflicts in runtime. That's why I chose to use a temporary file - to make the repository more abstract.

Reply:
The service would keep the local repo. And someone would change the configuration while switching the app on and off to a different remote. The service would keep the local repo and boom - merge conflicts in runtime. No. I would rather wait for `git clone` on every start.